python_modules/dagster/dagster/components/core/defs_module.py

```diff
@@ -200,12 +215,45 @@ def get_component(cls, context) -> Component:
             )
 
 
-@dataclass
+ENV_VAR_REGEX = re.compile(r"\{\{\s*env\(\s*['\"]([^'\"]+)['\"]\)\s*\}\}")
+
+
+def get_used_env_vars(data_structure: Union[Mapping[str, Any], Sequence[Any], Any]) -> set[str]:
+    if isinstance(data_structure, Mapping):
+        return set.union(set(), *(get_used_env_vars(value) for value in data_structure.values()))
+    elif isinstance(data_structure, str):
+        return set(ENV_VAR_REGEX.findall(data_structure))
+    elif isinstance(data_structure, Sequence):
+        return set.union(set(), *(get_used_env_vars(item) for item in data_structure))
+    return set()
```

Comment:
hmmm I am skeptical this is robust enough to rely on it to this degree. this is more like an extremely strong heuristic.

Perhaps if we stubbed out everything in scope and evaluated the template string we could rely on it more